### PR TITLE
Restructuring according to changed request/response format

### PR DESCRIPTION
**Pull request by stivni**
Two major things in request/response format changed:

- top level `data` attribute instead of _items
- `type` property on every reference to another object, both in read and write

IN SCOPE:
structure

OUT OF SCOPE:
type names
all other stuff!

### CODE REVIEW

File: apiary.apib [hunk 1]
```diff
@@ -1752,22 +1752,60 @@ Get a list of tracked time.
...
+                + topic (object)
+                    + type: `topic` (string)
+                    + id: `456` (string)
+                + started_on: `2017-04-26` (string)
+                + started_at: `2017-04-26T10:01:49+00:00` (string, nullable)
+                + ended_at: `2017-04-26T11:01:49+00:00` (string, nullable)
+                + duration: 3600 (number) - Tracked time in seconds
+                + description: `Timer description` (string)
+                + subject (object)
+                    + type: `project` (enum[string], required)
```
Comment (mrtus) [inline apiary.apib]: Q: This shouldn't be required for `read` operations right?

File: apiary.apib [hunk 2]
```diff
@@ -2349,24 +2349,29 @@ Create a draft credit note.
         + for_attention_of (object, optional)
             + name (string, optional)
             + contact_id (string, optional)
```
Comment (mrtus) [inline apiary.apib]: B: This should be
```
+ contact (object, optional)
    + type: `contact` (string, required)
    + id: `5` (string, required)
```

File: apiary.apib [hunk 3]
```diff
@@ -1440,10 +1445,15 @@ Update a calendar event.
...
         + description (string, optional)
-        + creator_id: 1 (string, optional)
-        + todo_id: 1 (string, optional)
-        + activity_type_id: 1 (string, optional)
-        + topic_id: 1 (string, optional)
+        + todo (object, optional)
+            + type: `todo` (string, required)
+            + id: 1 (string, required)
+        + activity_type (object, required)
+            + type: `activity_type` (string, required)
```
[the remark this answers is not in the document]
Comment (stivni) [reply apiary.apib]: yes, but I'll bring consistency in that in a later PR

File: apiary.apib (hunk 1 above)
Comment (stivni) [reply apiary.apib]: doesn't really matter, but will fix

File: apiary.apib (hunk 2 above)
Comment (stivni) [reply apiary.apib]: good catch

File: src/02-crm/companies.apib [hunk 4]
```diff
@@ -113,20 +113,23 @@ Add a new company.
         + iban: `BE12 1234 1234 1234` (string)
         + bic: `BICBANK` (string)
         + language: `en` (string, optional)
-        + responsible_user_id (string, optional)
+        + responsible_user (object, optional)
+            + type: `User` (string)
+            + id (string, optional)
```
Comment (regularjack) [inline src/02-crm/companies.apib]: Q: Should the `id` be optional? I think if a `responsible_user` is passed than the `id` should be mandatory. This seems to be how we do it in for example `/deals.create`.

Applies to other places too.

File: src/03-activities/events.apib [hunk 5]
```diff
@@ -129,10 +128,15 @@ Create a new calendar event.
...
         + description (string, optional)
-        + creator_id: 1 (string, optional)
-        + todo_id: 1 (string, optional)
-        + activity_type_id: 1 (string, required)
-        + topic_id: 1 (string, optional)
+        + todo (object, optional)
+            + type: `todo` (string, required)
+            + id: 1 (string, required)
+        + activity_type (object, required)
+            + type: `activity_type` (string, required)
```
Comment (regularjack) [inline src/03-activities/events.apib]: Q: `activityType`?

Comment (stivni) [reply src/03-activities/events.apib]: correct, type naming will be fixed in a follow up PR

File: src/02-crm/contacts.apib [hunk 6]
```diff
@@ -115,37 +115,40 @@ Add a new contact.
...
+                    + M
+                    + F
+                    + U
+            + birthdate: `1989-08-19` (string, optional)
+            + remarks: `Met at expo` (string, optional)
+            + tags: expo, upsell (array[string], optional)
+            + custom_fields (array)
+                + (object)
+                    + field (object)
+                        + type: `custom_field` (string)
```
Comment (Steerlin) [inline src/02-crm/contacts.apib]: type value should be `customField` ?

File: src/datastructures.apib [hunk 7]
```diff
@@ -101,17 +107,17 @@
 + line_items (array)
     + (object)
         + product (object)
-            + _type: `Product` (string)
+            + type: `Product` (string)
```
Comment (Steerlin) [inline src/datastructures.apib]: Ledger without capital

File: src/datastructures.apib [hunk 8]
```diff
@@ -101,17 +107,17 @@
...
             + id: `1` (string)
         + quantity: `3` (number)
         + description: `An awesome product` (string)
         + extended_description: `Additional information` (string)
         + unit_price (ReadPrice)
         + ledger (object)
-            + _type: `Ledger` (string) - ?????
+            + type: `Ledger` (string) - ?????
```
Comment (Steerlin) [inline src/datastructures.apib]: Ledger without capital

File: src/04-invoicing/subscriptions.apib [hunk 9]
```diff
@@ -10,53 +10,54 @@
...
+                + tax_inclusive (Money)
+                + payable (Money)
+                + taxes (array)
+                    + (object)
+                        + rate: `0.21` (number)
+                        + taxable (Money)
+                        + tax (Money)
+                + withheld_taxes (array)
+                    + (object)
+                        + type: `Trabajo`
```
Comment (Steerlin) [inline src/04-invoicing/subscriptions.apib]: should this type value be lowercased as well ?

File: src/04-invoicing/invoices.apib [hunk 10]
```diff
@@ -71,51 +71,52 @@ Get a single invoice.
...
+                + tax_inclusive (Money)
+                + payable (Money)
+                + taxes (array)
+                    + (object)
+                        + rate: `0.21` (number)
+                        + taxable (Money)
+                        + tax (Money)
+                + withheld_taxes (array)
+                    + (object)
+                        + type: `Trabajo`
```
Comment (Steerlin) [inline src/04-invoicing/invoices.apib]: lowercase?

File: src/02-crm/contacts.apib [hunk 11]
```diff
@@ -226,8 +229,8 @@ Link a contact to a company.
 
     + Attributes (object)
         + id: `123` (string, required)
-        + company (object)
-            + type: `company` (string, optional)
+        + company (object, required)
+            + type: `Company` (string, required)
```
Comment (Steerlin) [inline src/02-crm/contacts.apib]: type has capital

File: src/02-crm/contacts.apib [hunk 12]
```diff
@@ -242,8 +245,8 @@ Unlink a contact from a company.
...
```
Comment (Steerlin) [inline src/02-crm/contacts.apib]: type has capital

File: src/02-crm/companies.apib [hunk 13]
```diff
@@ -151,17 +155,17 @@ Update a company.
         + iban: `BE12 1234 1234 1234` (string)
         + bic: `BICBANK` (string)
         + language: `en` (string, optional)
-        + responsible_user (object)
-            + type: `user` (string)
-            + id: `456` (string)
+        + responsible_user (object, optional)
+            + type: `User` (string)
```
Comment (Steerlin) [inline src/02-crm/companies.apib]: type has capital

File: src/02-crm/companies.apib (hunk 4 above)
Comment (stivni) [reply src/02-crm/companies.apib]: :ok: by @stivni on 26/10/2017 13:06:59

File: src/datastructures.apib (hunk 7 above)
Comment (stivni) [reply src/datastructures.apib]: out of scope of this PR

 :ok: by @stivni on 26/10/2017 13:07:21

File: src/02-crm/contacts.apib (hunk 6 above)
Comment (stivni) [reply src/02-crm/contacts.apib]: out of scope of this PR

 :ok: by @stivni on 26/10/2017 13:07:30

File: src/04-invoicing/invoices.apib (hunk 10 above)
Comment (stivni) [reply src/04-invoicing/invoices.apib]: out of scope of this PR

 :ok: by @stivni on 26/10/2017 13:07:38

File: src/02-crm/contacts.apib (hunk 11 above)
Comment (stivni) [reply src/02-crm/contacts.apib]: out of scope of this PR

 :ok: by @stivni on 26/10/2017 13:07:44

File: src/02-crm/companies.apib (hunk 13 above)
Comment (stivni) [reply src/02-crm/companies.apib]: out of scope of this PR

 :ok: by @stivni on 26/10/2017 13:07:47

File: src/02-crm/contacts.apib (hunk 12 above)
Comment (stivni) [reply src/02-crm/contacts.apib]: out of scope of this PR

 :ok: by @stivni on 26/10/2017 13:07:52

File: src/04-invoicing/subscriptions.apib (hunk 9 above)
Comment (stivni) [reply src/04-invoicing/subscriptions.apib]: out of scope of this PR

 :ok: by @stivni on 26/10/2017 13:08:01

File: src/02-crm/deals.apib [hunk 14]
```diff
@@ -190,23 +180,15 @@ Update a deal.
...
-        + source (object, optional)
-            + type: `dealSource` (string, required)
-            + id: `1` (string, required)
-        + department (object, optional)
-            + type: `dealSource` (string, required)
-            + id: `1` (string, required)
-        + responsible_user (object, optional)
-            + type: `user` (string, required)
-            + id: `1` (string, required)
+        + source_id: `1` (string, required)
```
Comment (mrtus) [inline src/02-crm/deals.apib]: Q: This (among others) was optional before? What changed this?

Comment (stivni) [reply src/02-crm/deals.apib]: some were intentional, some were copy-pasta, checked & corrected them all (the errors)

 :ok: by @stivni on 26/10/2017 17:48:12

**Pull request merged**